ICU-20530 Fix header issues based on test/hdrtst

alonbl opened the pull request:
Remove explicit extern "C++" scope.

Fix issues detected by make -C test/hdrtst.

Run test/hdrtest during CI to detect future breakage.

Signed-off-by: Alon Bar-Lev <alon.barlev@gmail.com>

<!--
Thank you for your pull request.
Please see http://site.icu-project.org/processes/contribute for general
information on contributing to ICU.

You will be automatically asked to sign the contributors license before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/icu
- license: http://www.unicode.org/copyright.html#License
-->

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20530
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [x] Documentation is changed or added

alonbl commented:
This `makes -C test/hdrtst`  mostly work.

jefgen commented on icu4c/source/i18n/unicode/numberformatter.h:
```diff
@@ -112,7 +112,6 @@ class IntegerWidth;
 
 namespace impl {
 
-#ifndef U_HIDE_INTERNAL_API
```
Thanks for fixing this. Though I think this was fixed in the maint-64 branch during the 64.2 update. You might need to rebase and resolve conflicts.

alonbl replied:
yes, I just noticed, rebased.

jefgen commented on icu4c/source/i18n/unicode/numsys.h:
```diff
@@ -20,15 +20,12 @@
 
 #include "unicode/utypes.h"
 
-#ifndef U_HIDE_INTERNAL_API
```
I think this one might also have been fixed in the 64.2 update.
You might need to checkout master, fetch/pull, then rebase your branch on the updated master and resolve conflicts. (oh git.)

alonbl commented:
@jefgen, I believe this phase is complete, the headers assume global 'extern "C++"` global scope, removed the explicit scope, reduced conditionals by grouping C++ functions, added CI.

jefgen commented:
>     # Includes dependency checker.

Might be worth adding a comment that this also includes the hdrtst now too.

---
Refers to: .travis.yml:38 in 86bd2e9. [](commit_id = 86bd2e91c0d5eb438ceaaf26b080dd0d002a277b, deletion_comment = False)

jefgen commented on icu4c/source/common/unicode/uversion.h:
```diff
@@ -62,36 +62,29 @@ typedef uint8_t UVersionInfo[U_MAX_VERSION_LENGTH];
  * This is used to specify that the rest of the code uses the
  * public ICU C++ API namespace.
- * This is invoked by default; we recommend that you turn it off:
- * See the "Recommended Build Options" section of the ICU4C readme
```
Thanks for cleaning this up. IIRC, this was no longer set by default as of ICU 61.

[http://site.icu-project.org/download/61#TOC-Migration-Issues](http://site.icu-project.org/download/61#TOC-Migration-Issues)

alonbl replied:
Yes, I traced that down to be incorrect.

jefgen commented on icu4c/source/common/unicode/uversion.h:
```diff
@@ -133,12 +124,7 @@ typedef uint8_t UVersionInfo[U_MAX_VERSION_LENGTH];
         U_NAMESPACE_USE
 #   endif
-#else
-#   define U_NAMESPACE_BEGIN
```
I wonder if all usages of these macros are guarded by `U_SHOW_CPLUSPLUS_API` though?

alonbl replied:
Yes, the test/hdrtst check exactly that.

As we discussed I will create a separate PR to guard all C++ headers, but as far as we fixed, these macros are now being used only by C++ code under the `__cplusplus` conditional (or the `U_SHOW_CPLUSPLUS_API` for that matter) or under the assumption that the global scope is C++.

jefgen replied:
Thanks! :)

jefgen commented on icu4c/source/i18n/unicode/uspoof.h:
```diff
@@ -1525,8 +1302,243 @@ uspoof_getInclusionSet(UErrorCode *status);
+ */
+/**
+ * \cond
+ * Note: Doxygen is giving a bogus warning on this U_DEFINE_LOCAL_OPEN_POINTER.
```
What is the Doxygen warning that you are seeing?

alonbl replied:
This:

```
/tmp/icu/icu4c/source/i18n/unicode/uspoof.h:1347: warning: Member U_DEFINE_LOCAL_OPEN_POINTER(LocalUSpoofCheckerPointer, USpoofChecker, uspoof_close) (function) of file uspoof.h is not documented.
```

It is the same issue with other variables such as `LocalUSpoofCheckResultPointer` which already has this condition, I guess doxygen is confused from the complex macro.

jefgen replied:
Thanks!
It looks like this is how the warning was avoided in ICU-13810, so makes sense to use the same approach. Thanks for fixing the warning.

jefgen commented on .travis.yml:
```diff
@@ -45,14 +45,17 @@ matrix:
       script:
         - make -j2 check
         - ( cd ../icu4c/source/test/depstest && ./depstest.py ../../../../build/ )
+        - make install
```
I suppose you could cd to `test/hdrtst` first, then just do `make check`, but I don't think it really matters.

alonbl replied:
The test/hdrtest is testing installed headers, not in-tree headers, so we need to install the package and then test.

jefgen replied:
Sorry, I was confused.

alonbl commented:
> > ```
> > # Includes dependency checker.
> > ```
> 
> Might be worth adding a comment that this also includes the hdrtst now too.
> 
> Refers to: .travis.yml:38 in [86bd2e9](https://github.com/unicode-org/icu/commit/86bd2e91c0d5eb438ceaaf26b080dd0d002a277b). [](commit_id = [86bd2e9](https://github.com/unicode-org/icu/commit/86bd2e91c0d5eb438ceaaf26b080dd0d002a277b), deletion_comment = False)

Done

jefgen commented:
I added a note to the JIRA ticket, but also adding a comment here as well:
I'd like to get review from other ICU-TC member(s) before merging this PR.

markusicu commented:
Jeff, do you want to push the "rebase and merge" button?

jefgen commented:
Thanks Markus! Yes, I can take over merging this PR into master.

Thanks again @alonbl for the PR!
